File: sources/Android/android-communications/src/main/java/com/androidcommunications/polar/api/ble/BleLogger.java
package com.androidcommunications.polar.api.ble;

public class BleLogger {

    public interface BleLoggerInterface {
        void d(final String tag, final String msg);

        void e(final String tag, final String msg);

        void w(final String tag, final String msg);

        void i(final String tag, final String msg);
    }

    public static String byteArrayToHex(final byte[] a) {
        StringBuilder sb = new StringBuilder(a.length * 2);
        for (byte b : a) {
            sb.append(String.format("%02X ", b & 0xff));
        }
        return sb.toString();
    }

    private static final BleLogger instance = new BleLogger();
    private BleLoggerInterface bleLoggerInterface = null;
    private final Object mutex = new Object();

    public static void setLoggerInterface(BleLoggerInterface loggerInterface) {
        synchronized (instance.mutex) {
            instance.bleLoggerInterface = loggerInterface;
        }
    }

    public static void d_hex(final String tag, final String msg, final byte[] data) {
        synchronized (instance.mutex) {
            if (instance.bleLoggerInterface != null) {
                instance.bleLoggerInterface.d(tag, msg + " HEX: " + byteArrayToHex(data));
            }
        }
    }

    public static void d(final String tag, final String msg) {
        synchronized (instance.mutex) {
            if (instance.bleLoggerInterface != null) {
                instance.bleLoggerInterface.d(tag, msg);
            }
        }
    }

    public static void e(String tag, String msg) {
        synchronized (instance.mutex) {
            if (instance.bleLoggerInterface != null) {
                instance.bleLoggerInterface.e(tag, msg);
            }
        }
    }

    public static void w(String tag, String msg) {
        synchronized (instance.mutex) {
            if (instance.bleLoggerInterface != null) {
                instance.bleLoggerInterface.w(tag, msg);
            }
        }
    }

    public static void i(String tag, String msg) {
        synchronized (instance.mutex) {
            if (instance.bleLoggerInterface != null) {
                instance.bleLoggerInterface.i(tag, msg);
            }
        }
    }
}

File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/BleLogger.kt
package com.polar.androidcommunications.api.ble

import kotlin.experimental.and

class BleLogger {
    interface BleLoggerInterface {
        fun d(tag: String, msg: String)
        fun e(tag: String, msg: String)
        fun w(tag: String, msg: String)
        fun i(tag: String, msg: String)
        fun d_hex(tag: String, msg: String, data: ByteArray)
    }

    private var bleLoggerInterface: BleLoggerInterface? = null
    private val mutex = Any()

    companion object {

        @JvmStatic
        fun byteArrayToHex(a: ByteArray): String {
            val sb = StringBuilder(a.size * 2)
            for (b: Byte in a) {
                sb.append(String.format("%02X ", b and 0xff.toByte()))
            }
            return sb.toString()
        }

        private val instance = BleLogger()

        @JvmStatic
        fun setLoggerInterface(loggerInterface: BleLoggerInterface?) {
            synchronized(instance.mutex) { instance.bleLoggerInterface = loggerInterface }
        }

        @JvmStatic
        fun d_hex(tag: String, msg: String, data: ByteArray) {
            synchronized(instance.mutex) {
                instance.bleLoggerInterface?.d(tag, msg + " HEX: " + byteArrayToHex(data))
            }
        }

        @JvmStatic
        fun d(tag: String, msg: String) {
            synchronized(instance.mutex) {
                instance.bleLoggerInterface?.d(tag, msg)
            }
        }

        @JvmStatic
        fun e(tag: String, msg: String) {
            synchronized(instance.mutex) {
                instance.bleLoggerInterface?.e(tag, msg)
            }
        }

        @JvmStatic
        fun w(tag: String, msg: String) {
            synchronized(instance.mutex) {
                instance.bleLoggerInterface?.w(tag, msg)
            }
        }

        @JvmStatic
        fun i(tag: String, msg: String) {
            synchronized(instance.mutex) {
                instance.bleLoggerInterface?.i(tag, msg)
            }
        }
    }
}